### DISCUSSION TITLE
Best approach for external one-time-per-request log dump

### DISCUSSION THREAD

**Discussion by esimonetti** [brefphp/bref#1624] [Support] [answered]
Hello Bref community,

I am still in thinking mode about an idea.
I am planning to use Laravel + Bref + AWS + S3 (and a bunch of other things that are not relevant to this question) :)

What I would like to complete for this idea, is to provide some information (throughout the PHP/Laravel request) to S3, as a single S3 API call, at the end of the request.

Is there a way to easily do that, without using other services (eg: a database or a cookie) to:
1- "store" the info that needs to be logged that is collected on multiple methods, most likely in memory somewhere
2- detect when the request is about to return, and then send the info to S3

I would also like this to be resilient (e.g.: on failure perhaps queue the logging in SQS, again in a single message, for later... sending).

Thanks so much for the brainstorming help!

**Comment by mnapoli** [comment, marked as answer]
Hi! I'd write a class that stores information in memory, and then would send the data at the end of the request.

In Laravel, you can execute code at the end of an HTTP request using a middleware and the `terminate` method. Create a middleware using the following Artisan command:

```bash
php artisan make:middleware TerminateMiddleware
```

This will generate a new middleware class named `TerminateMiddleware` in the `app/Http/Middleware` directory. Open the file and modify it:

```php
<?php

namespace App\Http\Middleware;

use Closure;

class TerminateMiddleware
{
    public function handle($request, Closure $next)
    {
        return $next($request);
    }

    public function terminate($request, $response)
    {
        // Code to run at the end of the request
    }
}
```

Next, you need to register the middleware in the `app/Http/Kernel.php` file. Add the middleware class to the `$middleware` array:

```php
protected $middleware = [
    // Other middleware entries
    \App\Http\Middleware\TerminateMiddleware::class,
];
```

Now, the `terminate` method in the `TerminateMiddleware` class will be automatically called at the end of each HTTP request.

**Reply by esimonetti** [reply]
Wow @mnapoli amazing, thanks for the great insights! This could be good content for the guide as well!

Thanks again
